ldap/client/api/src/main/java/org/apache/directory/ldap/client/api/LdapNetworkConnection.java

```diff
@@ -599,102 +604,100 @@ public boolean connect() throws LdapException
+        {
+            Future<LdapNetworkConnectionFuture> completedFuture;
+            try
             {
-                for ( LdapNetworkConnectionFuture cf : connFutures )
-                {
-                    if ( !cf.getConnectFuture().isDone() )
-                    {
-                        continue;
-                    }
+                completedFuture = completionService.take();
+                futures.remove( completedFuture );
+            }
+            catch ( InterruptedException e )
+            {
+                LOG.info( "InterruptedException occurred when waiting for a future to complete" );
+                continue;
+            }
 
-                    connFutures.remove( cf );
+            LdapNetworkConnectionFuture result;
+            try
+            {
+                result = completedFuture.get();
+            }
+            catch ( InterruptedException e )
+            {
+                LOG.info( "InterruptedException occurred when fetching future results - Error: " + e.getMessage() );
+                continue;
+            }
+            catch ( ExecutionException e )
+            {
+                LOG.info( "ExecutionException occurred when fetching future results - Error: " + e.getMessage() );
+                continue;
+            }
 
-                    ConnectFuture futureResult = null;
-                    try
-                    {
-                        futureResult = cf.getConnectFuture().get();
-                    }
-                    catch ( InterruptedException e )
-                    {
-                        LOG.info( "InterruptedException occurred when fetching future results - Error: " + e.getMessage() );
-                        cf.getConnector().dispose();
-                        continue;
-                    }
-                    catch ( ExecutionException e )
-                    {
-                        LOG.info( "ExecutionException occurred when fetching future results - Error: " + e.getMessage() );
-                        cf.getConnector().dispose();
-                        continue;
-                    }
+            if ( result.getConnectFuture() == null || !result.getConnectFuture().isConnected() )
+            {
+                result.getConnector().dispose();
+                continue;
+            }
 
-                    if ( futureResult == null || !futureResult.isConnected() )
-                    {
-                        cf.getConnector().dispose();
-                        continue;
-                    }
+            connectionFuture = result.getConnectFuture();
+            connector = result.getConnector();
 
-                    connFuture = cf;
-                    connectionFuture = futureResult;
-                    setConnector( cf.getConnector() );
+            break;
+        }
 
-                    break breakaway;
-                }
+        for ( Future<LdapNetworkConnectionFuture> future : futures )
```

Comment:
No not really. We might break out of the first loop in case we find a successful connection. So we might not loop through all the futures in that loop. The futures we do get a result from in the first loop, we do remove them. So the second loop is a cleanup loop to close and remove all the remaining futures.